### PR TITLE
Inline function keyword

### LINKED ISSUE #84: Allow 'inline' keyword in declarations
For now, it should just be a no-op.

```c
static inline void f(); 
<stdin>:1:8: error: expected variable name or '(', got 'inline'
```

### PR DESCRIPTION
**Pull request by pythondude325**
Added the inline keyword and the proper error when used in inappropriate places.

Closes #84

### CODE REVIEW

File: src/data/mod.rs [hunk 1]
```diff
@@ -173,18 +174,27 @@ impl Qualifiers {
...
     };
     pub const CONST_VOLATILE: Qualifiers = Qualifiers {
         c_const: true,
         volatile: true,
+        inline: false,
+    };
+    pub const INLINE: Qualifiers = Qualifiers {
+        c_const: false,
+        volatile: false,
+        inline: true,
```
Comment (jyn514) [inline src/data/mod.rs]: You don't need to add this unless you use it somewhere, the other `const` qualifiers are just helpers for unit tests mostly.

File: src/parse/decl.rs [hunk 2]
```diff
@@ -139,12 +139,13 @@ impl<I: Iterator<Item = Lexeme>> Parser<I> {
...
         };
-        if symbol.ctype.is_function() && qualifiers != Qualifiers::NONE {
+        if symbol.ctype.is_function() && (qualifiers.c_const || qualifiers.volatile) {
             warn(
                 &format!("{} has no effect on function return type", qualifiers),
                 id.location,
             );
-            qualifiers = Qualifiers::NONE;
+            qualifiers.c_const = false;
+            qualifiers.volatile = false;
```
Comment (jyn514) [inline src/parse/decl.rs]: Good catch, I would have missed this!

File: src/parse/decl.rs [hunk 3]
```diff
@@ -979,6 +981,20 @@ impl<I: Iterator<Item = Lexeme>> Parser<I> {
...
+            ..
+        }) = prefix
+        {
+            // `inline` is allowed on function declarations
+        } else if qualifiers.inline {
+            self.semantic_err(
+                "`inline` is only allowed on function declarations",
+                self.last_location,
+            );
+        }
```
Comment (jyn514) [inline src/parse/decl.rs]: So I know I keep changing my mind, but I think this would actually fit better in `Declarator::parse_type`. That way you don't have to keep passing the `qualifiers` down, you can just check them once at the end. Sorry for the inconvenience!

File: src/parse/decl.rs [hunk 4]
```diff
@@ -209,7 +210,7 @@ impl<I: Iterator<Item = Lexeme>> Parser<I> {
         self.expect(Token::Comma)?;
         loop {
             let decl = self
-                .declarator(false)?
+                .declarator(false, first_qualifiers)?
                 .expect("declarator should return Some when called with allow_abstract: false");
             let location = decl.id().unwrap().location;
             let (id, ctype) = decl
```
Comment (jyn514) [inline src/parse/decl.rs]: This should also check that `inline` isn't set. From 6.7.4  Function specifiers, paragraph 2:

> Function specifiers shall be used only in the declaration of an identifier for a function.

File: src/parse/decl.rs (hunk 3 above)
Comment (jyn514) [reply src/parse/decl.rs]: Actually given all the other places you have to check maybe it would be better to just have each call site check it individually ...

File: src/parse/decl.rs [hunk 5]
```diff
@@ -835,7 +836,7 @@ impl<I: Iterator<Item = Lexeme>> Parser<I> {
             }
             let (sc, quals, param_type, _) = self.declaration_specifiers()?;
             // true: allow abstract_declarators
-            let declarator = self.declarator(true)?;
+            let declarator = self.declarator(true, quals)?;
```
Comment (jyn514) [inline src/parse/decl.rs]: Technically `int f(inline int g());` should be an error but I don't think we need to catch it until https://github.com/jyn514/rcc/issues/142 is fixed.

File: src/parse/decl.rs (hunk 3 above)
Comment (jyn514) [reply src/parse/decl.rs]: I changed my mind again :laughing:  I think this is fine - it lets us check if we're in a function and the parameter decayed to a pointer. Specifically I'd like to give a better error than clang here:

```c
int f(inline int g());
```

clang: `<stdin>:1:7: error: 'inline' can only appear on functions`
gcc: `<stdin>:1:18: warning: parameter ‘g’ declared ‘inline’`
what I'd like to see: `error: 'inline' can only appear on functions. help: function parameters decay to pointers`

So long story short your current code is fine :)

File: src/parse/decl.rs [hunk 6]
```diff
@@ -1895,7 +1895,7 @@ mod tests {
                 .unwrap()
                 .is_err()
         );
-        assert!(parse("typedef a inlint int;").unwrap().is_err());
+        assert!(parse("typedef a inline int;").unwrap().is_err());
```
Comment (jyn514) [inline src/parse/decl.rs]: Maybe also check that this is a semantic error, not a syntax one?
